common/src/main/java/io/netty/util/DefaultAttributeMap.java

```diff
@@ -17,186 +17,240 @@
+                return newAttribute;
             }
         }
     }
 
     @Override
     public <T> boolean hasAttr(AttributeKey<T> key) {
         ObjectUtil.checkNotNull(key, "key");
-        AtomicReferenceArray<DefaultAttribute<?>> attributes = this.attributes;
-        if (attributes == null) {
-            // no attribute exists
-            return false;
-        }
-
-        int i = index(key);
-        DefaultAttribute<?> head = attributes.get(i);
-        if (head == null) {
-            // No attribute exists which point to the bucket in which the head should be located
-            return false;
-        }
+        DefaultAttribute[] attributes = this.attributes;
+        return searchAttributeByKey(attributes, 0, attributes.length, key) >= 0;
+    }
 
-        // We need to synchronize on the head.
-        synchronized (head) {
-            // Start with head.next as the head itself does not store an attribute.
-            DefaultAttribute<?> curr = head.next;
-            while (curr != null) {
-                if (curr.key == key && !curr.removed) {
-                    return true;
-                }
-                curr = curr.next;
+    private <T> void removeAttributeIfMatch(AttributeKey<T> key, DefaultAttribute<T> value) {
+        for (;;) {
+            final DefaultAttribute[] attributes = this.attributes;
+            final int count = attributes.length;
+            final int index = searchAttributeByKey(attributes, 0, count, key);
+            if (index < 0) {
+                return;
+            }
+            final DefaultAttribute attribute = attributes[index];
+            assert attribute.key() == key;
+            if (attribute != value) {
+                return;
+            }
+            final int newCount = count - 1;
+            final DefaultAttribute[] newAttributes =
+                    newCount == 0? EMPTY_ATTRIBUTES : new DefaultAttribute[newCount];
+            // perform 2 bulk copies
+            System.arraycopy(attributes, 0, newAttributes, 0, index);
+            final int remaining = count - index - 1;
+            if (remaining > 0) {
+                System.arraycopy(attributes, index + 1, newAttributes, index, remaining);
+            }
+            if (ATTRIBUTES_UPDATER.compareAndSet(this, attributes, newAttributes)) {
+                return;
             }
-            return false;
         }
     }
 
-    private static int index(AttributeKey<?> key) {
-        return key.id() & MASK;
-    }
-
     @SuppressWarnings("serial")
-    private static final class DefaultAttribute<T> extends AtomicReference<T> implements Attribute<T> {
+    private static final class DefaultAttribute<T> implements Attribute<T> {
+
+        private static final AtomicReferenceFieldUpdater<DefaultAttribute, Object> VALUE_UPDATER =
+                AtomicReferenceFieldUpdater.newUpdater(DefaultAttribute.class, Object.class, "value");
```

Comment:
Just curious, why change from extending `AtomicReference`?

Reply:
Because the old semantic wasn't documenting how to manage a removed state and now it doesn't match anymore the atomic reference one. I have strengthened the new semantic to reduce the size of the default attribute instance ;)

Reply:
OK sure, though I think you could still have just overridden those and used `super.get()` etc? Only for a slight simplification i.e. avoid needing the ARFU. Fine to leave as-is too

Reply:
I have chosen to not do that because at the beginning I have already tried it and found that the whole thing was compiling fine but failing some tests because of the ARFU existing implementation: it tooks me a bit to fine the issue so I have thought was a kind of semantic code smell and fixing it was I have currently done :)